### PR TITLE
feat(ops): support np.argpartition

### PR DESCRIPTION
**Pull request by lpizzinidev**
Adds support for the [`numpy.argpartition`](https://numpy.org/doc/stable/reference/generated/numpy.argpartition.html) operation.

### CODE REVIEW

**Comment by lpizzinidev** [comment]
**Note**
This is still a work in progress as the TF implementation is failing unit tests.
Any feedback on how to adjust the PR is appreciated! 🙏

File: keras/src/backend/torch/numpy.py [hunk 1]
```diff
@@ -1608,3 +1608,23 @@ def select(condlist, choicelist, default=0):
     for c, v in reversed(list(zip(condlist, choicelist))):
         out = torch.where(c, v, out)
     return out
+
+
+def argpartition(x, kth, axis=-1):
+    x = convert_to_tensor(x)
```
Comment (fchollet) [inline keras/src/backend/torch/numpy.py:1614]: Should we cast to int32 here?

File: keras/src/backend/torch/numpy.py [hunk 2]
```diff
@@ -1608,3 +1608,23 @@ def select(condlist, choicelist, default=0):
...
+    x = torch.transpose(x, axis, -1)
+    bottom_ind = torch.topk(-x, kth + 1)[1]
+
+    def set_to_zero(a, i):
+        a[i] = 0
+        return a
+
+    for _ in range(x.dim() - 1):
+        set_to_zero = torch.vmap(set_to_zero)
+    proxy = set_to_zero(torch.ones(x.shape), bottom_ind)
```
Comment (fchollet) [inline keras/src/backend/torch/numpy.py:1625]: Should ones be int32?

File: keras/src/ops/numpy.py [hunk 3]
```diff
@@ -6303,3 +6304,49 @@ def select(condlist, choicelist, default=0):
...
+
+        dtype = "int32"
+        if backend.backend() in ("torch", "numpy"):
+            dtype = "int64"
+        return KerasTensor(x.shape, dtype=dtype)
+
+
+@keras_export(["keras.ops.argpartition", "keras.ops.numpy.argpartition"])
+def argpartition(x, kth, axis=-1):
+    """Performs an indirect partition along the given axis. It returns an array
```
Comment (fchollet) [inline keras/src/ops/numpy.py:6332]: Add 2 line breaks after "the given axis."

File: keras/src/ops/numpy.py [hunk 4]
```diff
@@ -6303,3 +6304,49 @@ def select(condlist, choicelist, default=0):
...
+    Args:
+        a: Array to sort.
+        kth: Element index to partition by.
+            The k-th element will be in its final sorted position and all
+            smaller elements will be moved before it and all larger elements
+            behind it. The order of all elements in the partitions is undefined.
+            If provided with a sequence of k-th it will partition all of them
+            into their sorted position at once.
+        axis: Axis along which to sort. The default is -1 (the last axis).
+            If None, the flattened array is used.
```
Comment (fchollet) [inline keras/src/ops/numpy.py:6345]: Backticks around `None`

File: keras/src/backend/tensorflow/numpy.py [hunk 5]
```diff
@@ -2430,3 +2430,29 @@ def correlate(x1, x2, mode="valid"):
 
 def select(condlist, choicelist, default=0):
     return tf.experimental.numpy.select(condlist, choicelist, default=default)
+
+
+def argpartition(x, kth, axis=-1):
+    x = convert_to_tensor(x)
+
+    if standardize_dtype(x.dtype) not in ["int32", "int64"]:
```
Comment (fchollet) [inline keras/src/backend/tensorflow/numpy.py:2438]: Use `x.dtype.is_integer`

File: keras/src/backend/tensorflow/numpy.py [hunk 6]
```diff
@@ -2430,3 +2430,29 @@ def correlate(x1, x2, mode="valid"):
...
+        x = tf.cast(x, tf.int32)
+
+    x = tf.experimental.numpy.swapaxes(x, axis, -1)
```
Comment (fchollet) [inline keras/src/backend/tensorflow/numpy.py:2441]: Use `swapaxes` (defined in this very file)

File: keras/src/backend/tensorflow/numpy.py [hunk 7]
```diff
@@ -2430,3 +2430,29 @@ def correlate(x1, x2, mode="valid"):
...
+    bottom_ind = tf.math.top_k(-x, kth + 1)[1]
+
+    def set_to_zero(args):
+        a, i = args
+        updates = tf.reshape(tf.zeros_like(i, dtype=tf.float32), [-1])
```
Comment (fchollet) [inline keras/src/backend/tensorflow/numpy.py:2446]: Why float32 here?

File: keras/src/backend/tensorflow/numpy.py [hunk 8]
```diff
@@ -2430,3 +2430,29 @@ def correlate(x1, x2, mode="valid"):
...
+    for _ in range(x.ndim - 1):
+        set_to_zero = tf.vectorized_map(
+            set_to_zero, (tf.ones(tf.shape(x)), bottom_ind)
+        )
+    proxy = set_to_zero((tf.ones(tf.shape(x)), bottom_ind))
+
+    top_ind = tf.math.top_k(proxy, tf.shape(x)[-1] - kth - 1)[1]
+
+    out = tf.concat([bottom_ind, top_ind], axis=x.ndim - 1)
+    return tf.experimental.numpy.swapaxes(out, -1, axis)
```
Comment (fchollet) [inline keras/src/backend/tensorflow/numpy.py:2458]: Likewise, use `swapaxes`

File: keras/src/backend/tensorflow/numpy.py [hunk 9]
```diff
@@ -2430,3 +2430,29 @@ def correlate(x1, x2, mode="valid"):
...
+    def set_to_zero(args):
+        a, i = args
+        updates = tf.reshape(tf.zeros_like(i, dtype=tf.float32), [-1])
+        return tf.tensor_scatter_nd_update(a, i, updates)
+
+    for _ in range(x.ndim - 1):
+        set_to_zero = tf.vectorized_map(
+            set_to_zero, (tf.ones(tf.shape(x)), bottom_ind)
+        )
+    proxy = set_to_zero((tf.ones(tf.shape(x)), bottom_ind))
```
Comment (fchollet) [inline keras/src/backend/tensorflow/numpy.py:2453]: This appears to be reimplementing the corresponding part of the JAX code that uses `vmap`. But how did you come with this exactly? I can't quite follow the logic.

**Review by fchollet — COMMENTED** [review]
Thanks for the PR!

**Comment by lpizzinidev** [comment]
@fchollet 
Thanks for the feedback! I'm working to fix the TF implementation.
Hopefully, this will be ready in the following days.

**Comment by lpizzinidev** [comment]
@fchollet 
Changed the approach a bit for TF, tests are green.
Please let me know if further changes are needed 💪

**Review by fchollet — COMMENTED** [review]
Thanks for the updates! Please also format the code by running `sh shell/format.sh`.

**Review by fchollet — COMMENTED** [review]
The torch unit tests fail -- it looks like we need to cast outputs to int32 for torch. https://github.com/keras-team/keras/actions/runs/8853599761/job/24314819196?pr=19588

**Review by fchollet — APPROVED** [review]
Thank you for the contribution! LGTM

**Comment by fchollet** [comment]
Torch tests seem to fail on GPU with device placement issues: https://btx.cloud.google.com/invocations/43f688ba-e1b0-4100-8725-1b7756581e52/targets/keras%2Fgithub%2Fubuntu%2Fgpu%2Ftorch%2Fcontinuous/log

**Comment by fchollet** [comment]
I think in the line `out = torch.cat([bottom_ind, top_ind], dim=x.dim() - 1)` we have one input on CPU and one on GPU. we should figure out why that happens and fix it.

**Comment by fchollet** [comment]
I fixed it. The issue was the use of `torch.ones` (which does not have automated device placement) instead of `ops.ones` (which does).

**Comment by lpizzinidev** [comment]
@fchollet 
Thanks for the fix and the explanation! 👍
Is there a public link for contributors to check on these post-merge builds? 🤔
